File: typescipt/src/compiler/cSharp/expressionStatementExceptions/fillFunctionExpressionStatementException.ts


export class FillFunctionExpressionStatementException extends IExpressionStatementException {
   public matches(expression: Expression): boolean {
     return expression is VariableDeclarationExpression assignmentExpression
        && assignmentExpression.Assignment is FunctionCallExpression functionCallExpression
        && functionCallExpression.ExpressionFunction is FillParametersFunction;
   }

   public callExpressionSyntax(expression: Expression, context: ICompileFunctionContext): Array<StatementSyntax> {
     if (!(expression is VariableDeclarationExpression assignmentExpression))
       throw new Error(`expression should be AssignmentExpression`);
     if (!(assignmentExpression.Assignment is FunctionCallExpression functionCallExpression))
       throw new Error(`assignmentExpression.Assignment should be FunctionCallExpression`);
     if (!(functionCallExpression.ExpressionFunction is FillParametersFunction fillParametersFunction))
       throw new Error(
         `functionCallExpression.ExpressionFunction should be FillParametersFunction`);

     return FillStatementSyntax(assignmentExpression.Name, fillParametersFunction.Type,
       fillParametersFunction.Mapping);
   }

   public static Array<StatementSyntax> FillStatementSyntax(string variableName, VariableType type,
     Array<Mapping> mappings) {
     if (variableName == null) throw new Error(nameof(variableName));
     if (type == null) throw new Error(nameof(type));
     if (mappings == null) throw new Error(nameof(mappings));

     let typeSyntax = Types.Syntax(type);

     let initialize = ObjectCreationExpression(Types.Syntax(type))
       .WithArgumentList(ArgumentList());

     let variable = VariableDeclarator(Identifier(variableName))
       .WithInitializer(EqualsValueClause(initialize));

     yield return LocalDeclarationStatement(
       VariableDeclaration(typeSyntax)
         .WithVariables(SingletonSeparatedList(variable)));

     foreach (let mapping in mappings) {
       let left = MemberAccessExpression(
         SyntaxKind.SimpleMemberAccessExpression,
         IdentifierName(variableName),
         IdentifierName(mapping.VariableName));

       let right = mapping.VariableSource == VariableSource.Code
         ? IdentifierName(mapping.VariableName)
         : mapping.VariableSource == VariableSource.Parameters
           ? (ExpressionSyntax)MemberAccessExpression(
             SyntaxKind.SimpleMemberAccessExpression,
             IdentifierName(LexyCodeConstants.ParameterVariable),
             IdentifierName(mapping.VariableName))
           : throw new Error(`Invalid source: ` + mapping.VariableSource);

       yield return ExpressionStatement(
         AssignmentExpression(SyntaxKind.SimpleAssignmentExpression, left, right));
     }
   }
}

File: src/Lexy.Compiler/Compiler/CSharp/ExpressionStatementExceptions/FillFunctionExpressionStatementException.cs
using System;
using System.Collections.Generic;
using Lexy.Compiler.Language;
using Lexy.Compiler.Language.Expressions;
using Lexy.Compiler.Language.Expressions.Functions;
using Lexy.Compiler.Language.VariableTypes;
using Microsoft.CodeAnalysis.CSharp;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using static Microsoft.CodeAnalysis.CSharp.SyntaxFactory;

namespace Lexy.Compiler.Compiler.CSharp.ExpressionStatementExceptions;

internal class FillFunctionExpressionStatementException : IExpressionStatementException
{
    public bool Matches(Expression expression)
    {
        return expression is VariableDeclarationExpression { Assignment: FunctionCallExpression functionCallExpression }
            && functionCallExpression is FillParametersFunction;
    }

    public IEnumerable<StatementSyntax> CallExpressionSyntax(Expression expression)
    {
        if (expression is not VariableDeclarationExpression assignmentExpression)
            throw new InvalidOperationException("expression should be AssignmentExpression");
        if (assignmentExpression.Assignment is not FunctionCallExpression functionCallExpression)
            throw new InvalidOperationException("assignmentExpression.Assignment should be FunctionCallExpression");
        if (functionCallExpression is not FillParametersFunction fillParametersFunction)
            throw new InvalidOperationException(
                "functionCallExpression.FunctionCallExpression should be FillParametersFunction");

        return FillStatementSyntax(assignmentExpression.Name, fillParametersFunction.Type,
            fillParametersFunction.Mapping);
    }

    public static IEnumerable<StatementSyntax> FillStatementSyntax(string variableName, VariableType type,
        IEnumerable<Mapping> mappings)
    {
        if (variableName == null) throw new ArgumentNullException(nameof(variableName));
        if (type == null) throw new ArgumentNullException(nameof(type));
        if (mappings == null) throw new ArgumentNullException(nameof(mappings));

        var typeSyntax = Types.Syntax(type);

        var initialize = ObjectCreationExpression(Types.Syntax(type))
            .WithArgumentList(ArgumentList());

        var variable = VariableDeclarator(Identifier(variableName))
            .WithInitializer(EqualsValueClause(initialize));

        yield return LocalDeclarationStatement(
            VariableDeclaration(typeSyntax)
                .WithVariables(SingletonSeparatedList(variable)));

        foreach (var mapping in mappings)
        {
            var left = MemberAccessExpression(
                SyntaxKind.SimpleMemberAccessExpression,
                IdentifierName(variableName),
                IdentifierName(mapping.VariableName));

            var right = mapping.VariableSource == VariableSource.Code
                ? IdentifierName(mapping.VariableName)
                : mapping.VariableSource == VariableSource.Parameters
                    ? (ExpressionSyntax)MemberAccessExpression(
                        SyntaxKind.SimpleMemberAccessExpression,
                        IdentifierName(LexyCodeConstants.ParameterVariable),
                        IdentifierName(mapping.VariableName))
                    : throw new InvalidOperationException("Invalid source: " + mapping.VariableSource);

            yield return ExpressionStatement(
                AssignmentExpression(SyntaxKind.SimpleAssignmentExpression, left, right));
        }
    }
}